From stable+bounces-206360-greg=kroah.com@vger.kernel.org Thu Jan  8 16:14:10 2026
From: Thadeu Lima de Souza Cascardo <cascardo@igalia.com>
Date: Thu,  8 Jan 2026 12:03:49 -0300
Subject: ext4: filesystems without casefold feature cannot be mounted with siphash
To: stable@vger.kernel.org
Cc: Gabriel Krisman Bertazi <krisman@suse.de>, Lizhi Xu <lizhi.xu@windriver.com>, syzbot+340581ba9dceb7e06fb3@syzkaller.appspotmail.com, Theodore Ts'o <tytso@mit.edu>, Thadeu Lima de Souza Cascardo <cascardo@igalia.com>
Message-ID: <20260108150350.3354622-1-cascardo@igalia.com>

From: Lizhi Xu <lizhi.xu@windriver.com>

commit 985b67cd86392310d9e9326de941c22fc9340eec upstream.

When mounting the ext4 filesystem, if the default hash version is set to
DX_HASH_SIPHASH but the casefold feature is not set, exit the mounting.

Reported-by: syzbot+340581ba9dceb7e06fb3@syzkaller.appspotmail.com
Signed-off-by: Lizhi Xu <lizhi.xu@windriver.com>
Link: https://patch.msgid.link/20240605012335.44086-1-lizhi.xu@windriver.com
Signed-off-by: Theodore Ts'o <tytso@mit.edu>
[cascardo: small conflict fixup]
Signed-off-by: Thadeu Lima de Souza Cascardo <cascardo@igalia.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>
---
 fs/ext4/super.c |    8 ++++++++
 1 file changed, 8 insertions(+)

--- a/fs/ext4/super.c
+++ b/fs/ext4/super.c
@@ -3632,6 +3632,14 @@ int ext4_feature_set_ok(struct super_blo
 	}
 #endif
 
+	if (EXT4_SB(sb)->s_es->s_def_hash_version == DX_HASH_SIPHASH &&
+	    !ext4_has_feature_casefold(sb)) {
+		ext4_msg(sb, KERN_ERR,
+			 "Filesystem without casefold feature cannot be "
+			 "mounted with siphash");
+		return 0;
+	}
+
 	if (readonly)
 		return 1;
 
